### PR TITLE
DateTime format support

### PR DESCRIPTION
**Pull request by VaishnaviParaskar**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -72,49 +72,7 @@
...
-import static org.apache.calcite.sql.SqlDateTimeFormat.POST_MERIDIAN_INDICATOR;
-import static org.apache.calcite.sql.SqlDateTimeFormat.POST_MERIDIAN_INDICATOR1;
-import static org.apache.calcite.sql.SqlDateTimeFormat.SECOND;
-import static org.apache.calcite.sql.SqlDateTimeFormat.TIMEZONE;
-import static org.apache.calcite.sql.SqlDateTimeFormat.TWENTYFOURHOUR;
-import static org.apache.calcite.sql.SqlDateTimeFormat.TWODIGITYEAR;
-import static org.apache.calcite.sql.SqlDateTimeFormat.YYMMDD;
-import static org.apache.calcite.sql.SqlDateTimeFormat.YYYYMM;
-import static org.apache.calcite.sql.SqlDateTimeFormat.YYYYMMDD;
+import static org.apache.calcite.sql.SqlDateTimeFormat.*;
```
Comment (pratik163) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:75]: resolve checkStyle issues, remove wildcard import

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 2]
```diff
@@ -1439,7 +1439,7 @@ protected String getDateTimeFormatString(
         startIndex = i + 1;
       }
     }
-    if (lastIndex > startIndex) {
+    if (lastIndex >= startIndex) {
```
Comment (pratik163) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java:1442]: checking for gt and eq will result in test failure for some scenarios. Kindly retest for all sscenarios

**Review by pratik163 — COMMENTED** [review]
Some changes required. Kindle check comments

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 1 above)
Comment (VaishnaviParaskar) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: done

**Comment by VaishnaviParaskar** [comment]
all test cases are passing

**Review by pratik163 — APPROVED** [review]

**Review by pratik163 — APPROVED** [review]

**Comment by vaibhavjain-dm** [comment]
There are some checks failing please check.

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 3]
```diff
@@ -1448,10 +1444,14 @@ protected String getDateTimeFormatString(
   private String getFinalFormat(
       List<String> dateTimeTokens, List<Character> separators,
       Map<SqlDateTimeFormat, String> dateTimeFormatMap) {
+    Pattern numberRegex = Pattern.compile("^[0-9]*$");
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java:1447]: Use isNumeric method

**Comment by vaibhavjain-dm** [comment]
Checks are failing due to your code please check.

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 4]
```diff
@@ -1445,16 +1448,54 @@ protected String getDateTimeFormatString(
     return new Pair<>(dateTimeTokens, separators);
   }
 
+  private Pair<List<String>, List<Character>> getDateTimeTokensIfNotFound(
+      String standardDateFormat) {
+    List<String> dateTimeTokens = new ArrayList<>();
+    List<Character> separators = new ArrayList<>();
+    standardDateFormat = standardDateFormat.replace("''","'" );
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java:1455]: As discussed remove this as we are already handling in BigQuerySqlDialect.
Also check in other places as well.

Comment (VaishnaviParaskar) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 5]
```diff
@@ -148,15 +151,11 @@
 
   /**
    * Valid Date Time Separators.
+   * '@' is a standard Format added at mig side in case if we don't have any separators
+   * between two tokens which makes us easy to getFinalFormat
```
Comment (kk19ag) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java:155]: also add that these separators should be selected such that they are never a part of any pattern in SqlDateTimeFormat.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 6]
```diff
@@ -1078,9 +1092,18 @@ private void unparseRandomfunction(SqlWriter writer, SqlCall call, int leftPrec,
   @Override protected String getDateTimeFormatString(
       String standardDateFormat, Map<SqlDateTimeFormat, String> dateTimeFormatMap) {
     String dateTimeFormat = super.getDateTimeFormatString(standardDateFormat, dateTimeFormatMap);
-    return dateTimeFormat
+    dateTimeFormat = dateTimeFormat
         .replace("%Y-%m-%d", "%F")
-        .replace("%S.", "%E");
+        .replace("'", "");
```
Comment (kk19ag) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:1097]: Can you please create a JIRA for this replacement? we should test this more.

**Pull request merged**